src/seqence_similarity.py

```diff
@@ -0,0 +1,173 @@
+from typing import Callable
+
+import numpy as np
+import pandas as pd
+from scipy.stats import spearmanr
+
+from utils import load_records
+
+
+def preprocess_record(record: pd.DataFrame) -> pd.DataFrame:
+    """
+    Feature engineering some attributes than will help in identifying sequence
+
+    Args:
+        record (pd.DataFrame): single record with columns [start, end, pitch and velocity]
+
+    Returns:
+        pd.DataFrame: record with additional feature engineered columns [duration, step, absolute pitch]
+    """
+
+    # getting note duration
+    record["duration"] = record["end"] - record["start"]
+
+    # getting note step: distance between start of the note and start of previous note
+    record["step"] = record["start"].diff(periods=1)
+    # filling na for first value
+    record["step"].fillna(0)
+
+    # getting absolute note pitch
+    record["absolute_pitch"] = record["pitch"] % 12
+
+    return record
+
+
+def extract_relevant_attributes(df: pd.DataFrame) -> pd.DataFrame:
+    # extracting relevant attributes for sequence identification
+    # start and end are absolute values in context of the record so instead of them duration and step are used
+    return df[["pitch", "velocity", "duration", "step", "absolute_pitch"]]
```

Comment:
For comparing similarity I used attributes: ["pitch", "velocity", "duration", "step", "absolute_pitch"]. I think results could be better if instead of featured engineered attributes (especially I mean "pitch") I used some kind of learned embedding, but it would mean training some neural network.